builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1048,389 @@ func (b *backend) pathLoginRenew(
+	}
+
+	endpoint := "https://sts.amazonaws.com"
+
+	if config != nil {
+		if config.IAMServerIdHeaderValue != "" {
+			err = validateVaultHeaderValue(headers, parsedUrl, config.IAMServerIdHeaderValue)
+			if err != nil {
+				return logical.ErrorResponse(fmt.Sprintf("error validating %s header: %s", iamServerIdHeader, err)), nil
+			}
+		}
+		if config.STSEndpoint != "" {
+			endpoint = config.Endpoint
+		}
+	}
+
+	clientArn, accountID, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, sessionName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error parsing arn: %s", err)), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if !strutil.StrListContains(roleEntry.AllowedAuthTypes, iamAuthType) {
+		return logical.ErrorResponse(fmt.Sprintf("auth method iam not allowed for role %s", roleName)), nil
+	}
+
+	// The role creation should ensure that either we're inferring this is an EC2 instance
+	// or that we're binding an ARN
+	if roleEntry.BoundIamPrincipalARN != "" && roleEntry.BoundIamPrincipalARN != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal %q does not belong to the role %q", clientArn, roleName)), nil
+	}
+
+	policies := roleEntry.Policies
+	rTagMaxTTL := time.Duration(0)
+
+	inferredEntityType := ""
+	inferredEntityId := ""
+	if roleEntry.RoleInferredType == ec2EntityType {
+		instance, err := b.validateInstance(req.Storage, sessionName, roleEntry.InferredAWSRegion, accountID)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("failed to verify %s as a valid EC2 instance in region %s", sessionName, roleEntry.InferredAWSRegion)), nil
+		}
+
+		// build a fake identity doc to pass on metadata about the instance to verifyInstanceMeetsRoleRequirements
+		identityDoc := &identityDocument{
+			Tags:        nil, // Don't really need the tags, so not doing the work of converting them from Instance.Tags to identityDocument.Tags
+			InstanceID:  *instance.InstanceId,
+			AmiID:       *instance.ImageId,
+			AccountID:   accountID,
+			Region:      roleEntry.InferredAWSRegion,
+			PendingTime: instance.LaunchTime.Format(time.RFC3339),
+		}
+
+		roleTagResp, validationError, err := b.verifyInstanceMeetsRoleRequirements(req.Storage, instance, roleEntry, roleName, identityDoc)
+		if err != nil {
+			return nil, err
+		}
+		if validationError != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating instance: %s", validationError)), nil
+		}
+
+		if roleTagResp != nil {
```

Comment:
Yes; I view `disallow_reauthentication` as something that should only apply to the `ec2` authentication method (given its relatively static identity material, the instance identity document). And so, similarly, I view the `storedIdentity` as a construct for protecting against reply attacks of a discovered instance identity document that wouldn't need to be replicated when using the `iam` auth method given the rotating nature of those credentials.

RE: "I am inclined towards thinking that the inference of ec2_instance should only mean that we could apply the bounds that are applicable to an ec2 instance and nothing else." I largely agree with that, and that has been how I've tried to write this. But, RE: "We shouldn't allow role tag login from this code flow" I don't understand what you're saying. The use of inferencing means we've already decided to assume that the authenticated entity is an EC2 instance. And because role tags apply to EC2 instances, they should likewise apply to inferred EC2 instances. FWIW, I view role tagging as an authorization mechanism rather than an authentication mechanism (please correct me if I'm wrong!), and the code flow makes sure it is applicable only to EC2 instances.

Anyway, if I'm misunderstanding anything you're saying, please let me know!